devices/uart-gr716/uart.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * GR716 Serial driver
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Leczkowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <devices/devs.h>
+
+
+#define BUFFER_SIZE 0x200
+
+/* UART control bits */
+#define RX_FIFO_INT (1 << 10)
+#define TX_FIFO_INT (1 << 9)
+#define PARITY_EN   (1 << 5)
+#define TX_INT      (1 << 3)
+#define RX_INT      (1 << 2)
+#define TX_EN       (1 << 1)
+#define RX_EN       (1 << 0)
+
+/* UART status bits */
+#define RX_FIFO_FULL  (1 << 10)
+#define TX_FIFO_FULL  (1 << 9)
+#define TX_FIFO_EMPTY (1 << 2)
+#define TX_SR_EMPTY   (1 << 1)
+#define DATA_READY    (1 << 0)
+
+
+/* UART */
+enum {
+	uart_data,   /* Data register           : 0x00 */
+	uart_status, /* Status register         : 0x04 */
+	uart_ctrl,   /* Control register        : 0x08 */
+	uart_scaler, /* Scaler reload register  : 0x0C */
+	uart_dbg     /* FIFO debug register     : 0x10 */
+};
+
+
+typedef struct {
+	volatile u32 *base;
+	unsigned int irq;
+	u16 clk;
+
+	u8 dataRx[BUFFER_SIZE];
+	cbuffer_t cbuffRx;
+} uart_t;
+
+
+struct {
+	uart_t uarts[UART_MAX_CNT];
+} uart_common;
+
+
+const struct {
+	vu32 *base;
+	unsigned int irq;
+	unsigned int txPin;
+	unsigned int rxPin;
+} info[UART_MAX_CNT] = {
+	{ UART0_BASE, UART0_IRQ, UART0_TX, UART0_RX },
+	{ UART1_BASE, UART1_IRQ, UART1_TX, UART1_RX },
+	{ UART2_BASE, UART2_IRQ, UART2_TX, UART2_RX },
+	{ UART3_BASE, UART3_IRQ, UART3_TX, UART3_RX },
+	{ UART4_BASE, UART4_IRQ, UART4_TX, UART4_RX },
+	{ UART5_BASE, UART5_IRQ, UART5_TX, UART5_RX }
+};
+
+
+static inline void uart_rxData(uart_t *uart)
+{
+	char c;
+	/* Keep getting data until rx fifo is not empty */
+	while ((*(uart->base + uart_status) & DATA_READY) != 0) {
+		c = *(uart->base + uart_data) & 0xff;
+		lib_cbufWrite(&uart->cbuffRx, &c, 1);
+	}
+}
+
+
+static inline void uart_txData(uart_t *uart, const void *buff, size_t len)
+{
+	size_t i;
+	const char *c = buff;
+
+	for (i = 0; i < len; i++) {
+		/* Fill until tx fifo is not full */
+		while ((*(uart->base + uart_status) & TX_FIFO_FULL) != 0) {
+		}
+		*(uart->base + uart_data) = c[i];
+	}
+}
+
+
+static int uart_irqHandler(unsigned int n, void *data)
+{
+	uart_t *uart = (uart_t *)data;
+	u32 status = *(uart->base + uart_status);
+
+	if ((status & DATA_READY) != 0) {
+		uart_rxData(uart);
+	}
+
+	return 0;
+}
+
+
+static int uart_setPin(u8 pin)
+{
+	io_cfg_t uartCfg;
+	uartCfg.opt = 0x1;
+	uartCfg.pin = pin;
+	uartCfg.pullup = 0;
+	uartCfg.pulldn = 0;
+
+	switch (pin) {
+		case UART0_TX:
+		case UART1_TX:
+			/* pins used for SRAM */
+			return 0;
```

Comment:
We don't configure the pin, better return error
```suggestion
			return -1;
```